packages/gatsby-dev-cli/src/watch.js

```diff
@@ -3,15 +3,71 @@ const _ = require(`lodash`)
 const fs = require(`fs-extra`)
 const path = require(`path`)
 
+const {
+  publishPackagesLocallyAndInstall,
+} = require(`./local-npm-registry/verdaccio`)
+
+const { checkDepsChanges } = require(`./local-npm-registry/check-deps-changes`)
+const {
+  getDependantPackages,
+} = require(`./local-npm-registry/get-dependant-packages`)
+const { promisifiedSpawn } = require(`./local-npm-registry/utils`)
+
 let numCopied = 0
 
 const quit = () => {
   console.log(`Copied ${numCopied} files`)
   process.exit()
 }
 
-const copyPath = (oldPath, newPath, quiet) =>
-  new Promise((resolve, reject) => {
+const traversePackagesDeps = ({
+  root,
+  packages,
+  monoRepoPackages,
+  seenPackages = [...packages],
+  depTree = {},
+}) => {
+  packages.forEach(p => {
+    const pkgJson = require(path.join(root, `packages`, p, `package.json`))
+
+    const fromMonoRepo = _.intersection(
+      Object.keys({ ...pkgJson.dependencies }),
+      monoRepoPackages
+    )
+
+    fromMonoRepo.forEach(pkgName => {
+      depTree[pkgName] = (depTree[pkgName] || new Set()).add(p)
+    })
+
+    // only traverse not yet seen packages to avoid infinite loops
+    const newPackages = _.difference(fromMonoRepo, seenPackages)
+
+    if (newPackages.length) {
+      newPackages.forEach(depFromMonorepo => {
+        seenPackages.push(depFromMonorepo)
+      })
+
+      traversePackagesDeps({
```

Comment:
`seenPackages` is used to generate `newPackages`:
```
// only traverse not yet seen packages to avoid infinite loops
const newPackages = _.difference(fromMonoRepo, seenPackages)
```

`depTree` is used later (outside of this function) to see what other potential packages would need to be published - i.e. if we change dependency in `gatsby-cli` - checking `depTree` will know that `gatsby` package need to be published as well.

Does this explain it? I should add some comments there.